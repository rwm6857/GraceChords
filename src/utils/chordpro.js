export function parseChordPro(text){
  const lines = text.replace(/\r\n/g,'\n').split('\n'); const meta={}; const blocks=[]; let current={section:'',lines:[]}
  const metaRe=/^\{\s*([^:}]+)\s*:\s*([^}]*)\s*\}\s*$/
  for(const raw of lines){
    const m=raw.match(metaRe); if(m){ meta[m[1].trim().toLowerCase()] = m[2].trim(); continue }
    if(raw.trim()===''){ if(current.lines.length) blocks.push(current); current={section:'',lines:[]}; continue }
    const secMatch=raw.match(/^\s*(?:##?\s*|\[)([^#\]]+?)(?:\])?\s*$/);
    if(secMatch){
      const tag=secMatch[1].trim();
      const bracketOnly=/^\s*\[[^\]]+\]\s*$/.test(raw);
      const chordLike=/^[A-G][#b]?(?:m|maj|min|dim|aug|sus|add)?\d*(?:\/[A-G][#b]?)?$/i.test(tag);
      if(!raw.includes('[') || (bracketOnly && !chordLike)){
        if(current.lines.length) blocks.push(current); current={section:tag,lines:[]}; continue
      }
    }
    const { plain, chords } = extractChords(raw); current.lines.push({ text: plain, chords })
  } if(current.lines.length) blocks.push(current); return { meta, blocks }
}
export function extractChords(line){ let plain=''; const chords=[]; let i=0; while(i<line.length){ if(line[i]==='['){ const j=line.indexOf(']', i+1); if(j!==-1){ const sym=line.slice(i+1,j).trim(); chords.push({ sym, index: plain.length }); i=j+1; continue } } plain+=line[i]; i++ } return { plain, chords } }
export function makeMonospaceChordLine(plain, chordPositions){ if(!chordPositions?.length) return ''; let out=''; let cursor=0; for(const c of chordPositions){ const pad=Math.max(0, c.index - cursor); out += ' '.repeat(pad) + c.sym; cursor = c.index + c.sym.length } return out }
export const KEYS=['C','C#','D','D#','E','F','F#','G','G#','A','A#','B']; const FLAT={'Db':'C#','Eb':'D#','Gb':'F#','Ab':'G#','Bb':'A#'}; function norm(n){ return FLAT[n] || n }
export function stepsBetween(fromKey, toKey){ if(!fromKey||!toKey) return 0; const a=KEYS.indexOf(norm(fromKey)); const b=KEYS.indexOf(norm(toKey)); if(a===-1||b===-1) return 0; return (b-a+12)%12 }
export function transposeSym(sym, steps){ if(sym.includes('/')){ const [r,b]=sym.split('/'); return transposeSym(r,steps)+'/'+transposeSym(b,steps) } const m=sym.match(/^([A-G][#b]?)(.*)$/); if(!m) return sym; const i=KEYS.indexOf(norm(m[1])); if(i===-1) return sym; return KEYS[(i+steps+12)%12] + (m[2]||'') }
