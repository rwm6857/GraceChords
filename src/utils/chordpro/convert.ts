import { parseChordProOrLegacy } from './parser'
import { serializeChordPro } from './serialize'
import type { SongDoc } from './types'

export type MetaExtras = {
  country?: string
  tags?: string[] | string
  youtube?: string
  mp3?: string
  pptx?: string
}

function slugifyUnderscore(s: string){
  return (s || '').toLowerCase().replace(/[^\w]+/g,'_').replace(/_+/g,'_').replace(/^_|_$/g,'')
}

function stripDecorations(title: string){
  return title
    .replace(/\((israeli|iranian|hindi|arabic|.*?\blanguage\b)\)/ig,'')
    .replace(/\bkey\s+of\s+[A-G][#b]?m?\b/ig,'')
    .replace(/\s{2,}/g,' ')
    .trim()
}

export function convertToCanonicalChordPro(raw: string, extras?: Partial<MetaExtras>){
  const doc: SongDoc = parseChordProOrLegacy(raw)
  const tags = Array.isArray(extras?.tags) ? extras?.tags.join(', ') : (extras?.tags || '')
  doc.meta = {
    title: stripDecorations(doc.meta?.title || ''),
    key: (doc.meta?.key || '').trim(),
    capo: doc.meta?.capo,
    meta: {
      ...(doc.meta?.meta || {}),
      ...(extras?.country ? { country: extras.country } : {}),
      ...(tags ? { tags } : {}),
      ...(extras?.youtube ? { youtube: extras.youtube } : {}),
      ...(extras?.mp3 ? { mp3: extras.mp3 } : {}),
      ...(extras?.pptx ? { pptx: extras.pptx } : {}),
    }
  }

  const text = serializeChordPro(doc, { useDirectives: true })
  const docTitle = stripDecorations(doc.meta.title || 'Untitled')
  return { text, docTitle, docKey: doc.meta.key }
}

export function suggestCanonicalFilename(title: string){
  return `${slugifyUnderscore(stripDecorations(title || 'untitled'))}.chordpro`
}
