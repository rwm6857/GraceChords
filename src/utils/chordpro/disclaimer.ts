import { getChordproCommentBlock, isDisclaimerEnabled } from '../../config/disclaimer'

const MARKER = '# --- DISCLAIMER (GraceChords) ---'

export function hasDisclaimerCommentBlock(text: string): boolean {
  return String(text || '').split(/\r?\n/).some(line => line.trim() === MARKER)
}

export function appendDisclaimerIfMissing(text: string): string {
  if (!isDisclaimerEnabled()) return text
  const safe = String(text || '')
  if (hasDisclaimerCommentBlock(safe)) return safe
  const trimmed = safe.replace(/[\s\n]*$/, '')
  return trimmed + '\n\n' + getChordproCommentBlock()
}

export default { hasDisclaimerCommentBlock, appendDisclaimerIfMissing }

