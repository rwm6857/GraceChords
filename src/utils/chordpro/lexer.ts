export type LexToken =
  | { type: 'directive'; raw: string }
  | { type: 'lyrics'; raw: string }
  | { type: 'blank' };

export function lexChordPro(input: string): LexToken[] {
  const lines = input.split(/\r?\n/);
  return lines.map((raw) => {
    const t = raw.trim();
    if (t.startsWith('{') && t.endsWith('}')) return { type: 'directive', raw };
    if (t.length === 0) return { type: 'blank' };
    return { type: 'lyrics', raw };
  });
}
