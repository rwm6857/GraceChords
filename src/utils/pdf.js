// import { jsPDF } from 'jspdf'
import { ensureFontsEmbedded } from './fonts'

/** -----------------------------------------------------------------------
 *  Helpers & adapters
 *  -------------------------------------------------------------------- */

async function newPDF() {
  const { jsPDF } = await import('jspdf')
  return new jsPDF({ unit:'pt', format:'letter' })
}
/** Normalize input to the shape your renderer already uses:
 *  { title, key, lyricsBlocks: [{ section?: string, lines: [{ plain, chordPositions:[{index,sym}] }] }] }
 */
function normalizeSongInput(input) {
  if (input?.lyricsBlocks) {
    return {
      title: input.title || 'Untitled',
      key: input.key || input.originalKey || 'C',
      lyricsBlocks: input.lyricsBlocks
    }
  }
  // Adapter for parsed structures like: { title, originalKey, blocks:[ {type:'section'|'line', header?, lyrics?, chords?:[{index,sym}]} ] }
  if (Array.isArray(input?.blocks)) {
    const out = []
    let cur = { lines: [] }
    for (const b of input.blocks) {
      if (b.type === 'section') {
        if (cur.lines.length) out.push(cur)
        cur = { section: b.header, lines: [] }
      } else if (b.type === 'line') {
        cur.lines.push({
          plain: b.lyrics || b.text || '',
          chordPositions: (b.chords || []).map(c => ({ index: c.index ?? 0, sym: c.sym }))
        })
      }
    }
    if (cur.lines.length) out.push(cur)
    return {
      title: input.title || 'Untitled',
      key: input.key || input.originalKey || 'C',
      lyricsBlocks: out
    }
  }
  // Fallback (empty)
  return {
    title: input?.title || 'Untitled',
    key: input?.key || input?.originalKey || 'C',
    lyricsBlocks: []
  }
}

/** Create a lyrics-width measurer bound to a jsPDF doc + font settings */
function makeLyricMeasurer(doc, lyricFamily, lyricPt) {
  return (text) => {
    // Ensure measurement matches how lyrics are drawn
    doc.setFont(lyricFamily, 'normal')
    doc.setFontSize(lyricPt)
    return doc.getTextWidth(text || '')
  }
}

/** -----------------------------------------------------------------------
 *  PURE LAYOUT (source of truth)
 *  Computes pages/columns/blocks and chord X offsets from lyrics widths.
 *  No drawing here.
 *  -------------------------------------------------------------------- */

const DEFAULT_LAYOUT_OPT = {
  lyricSizePt: 16,
  chordSizePt: 16,
  margin: 36,
  gutter: 18,
  headerOffsetY: 54, // space below header before content
  columns: 1,
  pageWidth: 612,     // Letter pt (jsPDF default ~ 612x792) – we’ll read from doc in runtime path
  pageHeight: 792,
  lyricFamily: 'Helvetica',
  chordFamily: 'Courier',
  title: 'Untitled',
  key: 'C'
}

/**
 * computeLayout(songIn, opt, measureLyric)
 * - songIn: { title, key, lyricsBlocks:[{section?, lines:[{plain, chordPositions:[{index,sym}]}]}] }
 * - opt: see DEFAULT_LAYOUT_OPT
 * - measureLyric: (text) => width in pt, MUST match lyrics font used for PDF
 *
 * Returns:
 * {
 *   pages: [
 *     { columns: [
 *       { x, yStart, blocks: [
 *           { type:'section', header:'VERSE' },
 *           { type:'line', lyrics:'...', chords:[{ x, sym }, ...] }
 *       ] }
 *     ] }
 *   ]
 * }
 */
export function computeLayout(songIn, opt = {}, measureLyric = (t)=>0) {
  const song = normalizeSongInput(songIn)
  const o = { ...DEFAULT_LAYOUT_OPT, ...opt }
  const lineGap = 4
  const sectionSize = Math.max(o.lyricSizePt + 2, 16)

  const margin = o.margin
  const pageH = o.pageHeight
  const contentW = o.pageWidth - margin * 2
  const colW = o.columns === 2 ? (contentW - o.gutter) / 2 : contentW

  // Header metrics (mirror draw path)
  const contentStartY = margin + o.headerOffsetY
  const contentBottomY = pageH - margin

  const pages = []
  let page = { columns: [] }
  pages.push(page)

  function makeColumns() {
    const firstCol = { x: margin, yStart: contentStartY, blocks: [] }
    page.columns.push(firstCol)
    if (o.columns === 2) {
      const secondCol = { x: margin + colW + o.gutter, yStart: contentStartY, blocks: [] }
      page.columns.push(secondCol)
    }
  }
  function newPage() {
    page = { columns: [] }
    pages.push(page)
  }
  if (page.columns.length === 0) makeColumns()

  let colIdx = 0
  let cursorY = contentStartY

  const curCol = () => page.columns[colIdx]
  const advanceColOrPage = () => {
    if (o.columns === 2 && colIdx === 0) {
      colIdx = 1
      cursorY = contentStartY
    } else {
      newPage()
      makeColumns()
      colIdx = 0
      cursorY = contentStartY
    }
  }

  const measureBlockHeight = (block, fromLine = 0, toLineExclusive = (block.lines?.length ?? 0)) => {
    let h = 0
    const hasHeader = !!block.section && fromLine === 0
    if (hasHeader) h += sectionSize + 4
    for (let i = fromLine; i < toLineExclusive; i++) {
      const ln = block.lines[i]
      if (ln?.chordPositions?.length) h += o.chordSizePt + lineGap / 2
      h += o.lyricSizePt + lineGap
    }
    return h + 4 // block bottom padding
  }

  // Emit helpers
  const pushSection = (col, header) => col.blocks.push({ type: 'section', header: String(header).toUpperCase() })
  const pushLine = (col, plain, chordPositions) => {
    const chords = (chordPositions || []).map(c => ({
      sym: c.sym,
      x: measureLyric(plain.slice(0, c.index || 0))
    }))
    col.blocks.push({ type: 'line', lyrics: plain, chords })
  }

  for (const block of song.lyricsBlocks) {
    const lines = block.lines || []
    let start = 0
    while (start < lines.length) {
      // Determine the largest slice [start, end) that can fit in current column
      // but keep header attached to its first line.
      const col = curCol()

      // If nothing has been placed yet from this block and header exists, include it in height
      const minSliceH = measureBlockHeight(block, start, Math.min(start + 1, lines.length))
      if (cursorY + minSliceH > contentBottomY) {
        // Not even the header+first line fits -> advance column/page
        advanceColOrPage()
        continue
      }

      // Try to fit as many lines as possible
      let end = start + 1
      while (end <= lines.length) {
        const h = measureBlockHeight(block, start, end)
        if (cursorY + h > contentBottomY) break
        end++
      }
      end = Math.max(end - 1, start + 1) // at least one line (with header if start==0)

      // Emit this slice
      if (start === 0 && block.section) {
        pushSection(col, block.section)
        cursorY += sectionSize + 4
      }
      for (let i = start; i < end; i++) {
        const ln = lines[i]
        const plain = ln.plain || ln.text || ''
        const chordPositions = ln.chordPositions || []

        if (chordPositions.length) {
          // reserve chord row height in cursorY, but we only store blocks (y computed at draw time)
          // chords x were measured using lyrics widths
        }
        pushLine(col, plain, chordPositions)
        if (chordPositions.length) cursorY += o.chordSizePt + lineGap / 2
        cursorY += o.lyricSizePt + lineGap
      }
      cursorY += 4 // bottom padding for this slice

      // If we didn’t consume the whole block, advance to next column/page and continue
      start = end
      if (start < lines.length) {
        advanceColOrPage()
      }
    }
  }

  return { pages }
}


/** -----------------------------------------------------------------------
 *  DRAWING (consumes computeLayout)
 *  -------------------------------------------------------------------- */

function drawSongIntoDoc(doc, songIn, opt) {
  const lFam = String(opt.lyricFamily || 'Helvetica')
  const cFam = String(opt.chordFamily || 'Courier')
  const pageW = doc.internal.pageSize.getWidth()
  const pageH = doc.internal.pageSize.getHeight()
  const o = {
    ...DEFAULT_LAYOUT_OPT,
    ...opt,
    pageWidth: pageW,
    pageHeight: pageH
  }

  const margin = o.margin
  const contentW = pageW - margin * 2
  const gutter = o.gutter
  const colW = o.columns === 2 ? (contentW - gutter) / 2 : contentW

  // Header
  const headerTitlePt = Math.max(22, o.lyricSizePt + 6)
  const headerKeyPt   = Math.max(12, o.lyricSizePt - 2)
  doc.setFont(lFam, 'bold');   doc.setFontSize(headerTitlePt)
  doc.text(o.title, margin, margin + 24)
  doc.setFont(lFam, 'italic'); doc.setFontSize(headerKeyPt)
  doc.text(`Key: ${o.key || '—'}`, margin, margin + 40)

  // Layout using the SAME measurement path
  const measure = makeLyricMeasurer(doc, lFam, o.lyricSizePt)
  const layout = computeLayout(songIn, o, measure)

  const lineGap = 4
  const sectionSize = Math.max(o.lyricSizePt + 2, 16)
  const contentStartY = margin + o.headerOffsetY

  // Iterate pages/columns/blocks and draw
  layout.pages.forEach((p, pIdx) => {
    if (pIdx > 0) doc.addPage()
    p.columns.forEach((col, cIdx) => {
      let x = col.x
      let y = contentStartY
      for (const b of col.blocks) {
        if (b.type === 'section') {
          doc.setFont(lFam, 'bold'); doc.setFontSize(sectionSize)
          doc.text(`[${b.header}]`, x, y)
          y += sectionSize + 4
        } else if (b.type === 'line') {
          // Draw chords at x + chord.x (precomputed from lyrics widths)
          if (b.chords?.length) {
            doc.setFont(cFam, 'bold'); doc.setFontSize(o.chordSizePt)
            for (const c of b.chords) {
              doc.text(c.sym, x + c.x, y)
            }
            y += o.chordSizePt + lineGap / 2
          }
          // Draw lyric line
          doc.setFont(lFam, 'normal'); doc.setFontSize(o.lyricSizePt)
          doc.text(b.lyrics, x, y)
          y += o.lyricSizePt + lineGap
        }
      }
    })
  })
}

/** -----------------------------------------------------------------------
 *  PUBLIC PDF APIS (unchanged behavior)
 *  -------------------------------------------------------------------- */

export async function songToPdfDoc(song, options){
  const doc = await newPDF()
  const opt = {
    lyricSizePt: Math.max(14, options?.lyricSizePt || 16),
    chordSizePt: Math.max(14, options?.chordSizePt || 16),
    columns: options?.columns || 1,
    title: options?.title || song.title,
    key: options?.key || song.key,
    margin: 36,
    lyricFamily: 'Helvetica',
    chordFamily: 'Courier'
  }
  try{
    const f = await ensureFontsEmbedded(doc)
    opt.lyricFamily = f.lyricFamily || opt.lyricFamily
    opt.chordFamily = f.chordFamily || opt.chordFamily
  }catch{}
  drawSongIntoDoc(doc, song, opt)
  return doc
}

export async function downloadSingleSongPdf(song, options) {
  // Decide columns like before, but use layout pages length instead of drawing twice.
  const probeDoc = new jsPDF({ unit: 'pt', format: 'letter' })
  const pageW = probeDoc.internal.pageSize.getWidth()
  const pageH = probeDoc.internal.pageSize.getHeight()
  const opt1 = {
    lyricSizePt: Math.max(14, options?.lyricSizePt || 16),
    chordSizePt: Math.max(14, options?.chordSizePt || 16),
    columns: 1,
    title: options?.title || (song.title || 'Untitled'),
    key: options?.key || (song.key || 'C'),
    margin: 36,
    lyricFamily: 'Helvetica',
    chordFamily: 'Courier',
    pageWidth: pageW,
    pageHeight: pageH
  }
  // Measuring with default fonts is fine for page count decision.
  const layout1 = computeLayout(normalizeSongInput(song), opt1, makeLyricMeasurer(probeDoc, opt1.lyricFamily, opt1.lyricSizePt))
  const needsTwoCols = layout1.pages.length > 1

  const finalDoc = await songToPdfDoc(song, { ...options, columns: needsTwoCols ? 2 : 1 })
  finalDoc.save(`${(song.title || 'Untitled').replace(/\s+/g, '_')}.pdf`)
}

export async function downloadMultiSongPdf(songs, options){
  const doc = await newPDF()
  const f = await ensureFontsEmbedded(doc)
  const baseOpt = {
    lyricSizePt: Math.max(14, options?.lyricSizePt || 16),
    chordSizePt: Math.max(14, options?.chordSizePt || 16),
    columns: 1,
    margin: 36,
    lyricFamily: f.lyricFamily || 'Helvetica',
    chordFamily: f.chordFamily || 'Courier'
  }
  let first = true
  for(const s of songs){
    if(!first) doc.addPage()
    first = false
    drawSongIntoDoc(
      doc,
      { title: s.title, key: s.key, lyricsBlocks: s.lyricsBlocks },
      { ...baseOpt, title: s.title, key: s.key }
    )
  }
  doc.save('GraceChords_Selection.pdf')
}
/** -----------------------------------------------------------------------
 *  TEST-ONLY METRICS (pure snapshot; no drawing)
 *  These mirror layout math exactly.
 *  -------------------------------------------------------------------- */
export function getLayoutMetrics(input, opts) {
  // Use a tiny jsPDF purely for measurement so we match production widths.
  const probe = new jsPDF({ unit: 'pt', format: 'letter' })
  const pageW = probe.internal.pageSize.getWidth()
  const pageH = probe.internal.pageSize.getHeight()
  const o = {
    ...DEFAULT_LAYOUT_OPT,
    ...opts,
    pageWidth: pageW,
    pageHeight: pageH
  }
  // We do NOT embed fonts during tests by default. Helvetica metrics are deterministic and sufficient to catch algorithm regressions.
  const measure = makeLyricMeasurer(probe, o.lyricFamily, o.lyricSizePt)
  const layout = computeLayout(input, o, measure)

  // Normalize for snapshots (round chord x to 2 decimals)
  return layout.pages.map((page, pIdx) => ({
    p: pIdx,
    cols: page.columns.map((col, cIdx) => ({
      c: cIdx,
      blocks: col.blocks.map(b => ({
        t: b.type,
        h: b.type === 'section' ? b.header || null : null,
        line: b.type === 'line'
          ? {
              text: b.lyrics,
              chords: (b.chords || []).map(ch => ({ x: Number(ch.x.toFixed(2)), s: ch.sym }))
            }
          : null
      }))
    }))
  }))
}
