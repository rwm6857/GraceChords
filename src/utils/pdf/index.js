import { ensureFontsEmbedded } from './fonts'
import { planSongLayout, chooseBestLayout, normalizeSongInput, DEFAULT_LAYOUT_OPT } from './pdfLayout'
import { makeMeasure } from './measure'

// Debug switch: open DevTools and run localStorage.setItem('pdfDebug','1') to see guides
const PDF_DEBUG = typeof window !== 'undefined'
  && (() => { try { return localStorage.getItem('pdfDebug') === '1' } catch { return false } })()

/* -----------------------------------------------------------
 * Lazy jsPDF
 * --------------------------------------------------------- */
async function newPDF() {
  const { jsPDF } = await import('jspdf')
  return new jsPDF({ unit: 'pt', format: 'letter' })
}

/* -----------------------------------------------------------
 * Planner wrapper bound to jsPDF doc/fonts
 * --------------------------------------------------------- */
function planWithDoc(doc, song, baseOpt) {
  const pageW = doc.internal.pageSize.getWidth()
  const pageH = doc.internal.pageSize.getHeight()
  const oBase = { ...DEFAULT_LAYOUT_OPT, ...baseOpt, pageWidth: pageW, pageHeight: pageH }
  const makeLyric = makeMeasure(doc, oBase.lyricFamily, 'normal')
  const makeChord = makeMeasure(doc, oBase.chordFamily, 'bold')
  return chooseBestLayout(song, oBase, makeLyric, makeChord)
}

/* -----------------------------------------------------------
 * DRAWING (consumes planned layout)
 * --------------------------------------------------------- */
function drawPlannedSong(doc, plan, { title, key }) {
  const lFam = String(plan.lyricFamily || 'Helvetica')
  const cFam = String(plan.chordFamily || 'Courier')
  const pageW = doc.internal.pageSize.getWidth()
  const pageH = doc.internal.pageSize.getHeight()

  const margin = plan.margin
  const headerTitlePt = Math.max(22, plan.lyricSizePt + 6)
  const headerKeyPt   = Math.max(12, plan.lyricSizePt - 2)
  const lineGap = 4
  const sectionSize = plan.lyricSizePt
  const sectionTopPad = Math.round(plan.lyricSizePt * 0.85)
  const contentStartY = margin + plan.headerOffsetY
  const contentW = pageW - margin * 2
  const colW = plan.columns === 2 ? (contentW - plan.gutter) / 2 : contentW

  plan.layout.pages.forEach((p, pIdx) => {
    if (pIdx > 0) doc.addPage()
    // Header
    doc.setFont(lFam, 'bold');   doc.setFontSize(headerTitlePt)
    doc.text(title, margin, margin + 24)
    doc.setFont(lFam, 'italic'); doc.setFontSize(headerKeyPt)
    doc.text(`Key: ${key || '—'}`, margin, margin + 40)

    if (PDF_DEBUG) {
      doc.setDrawColor(180)
      doc.rect(margin, contentStartY, colW, pageH - margin - contentStartY)
      if (plan.columns === 2) {
        doc.rect(margin + colW + plan.gutter, contentStartY, colW, pageH - margin - contentStartY)
      }
      doc.setFont(lFam, 'normal'); doc.setFontSize(9)
      doc.text(
        `Plan: ${plan.columns} col • size ${plan.lyricSizePt}pt • singlePage=${plan.layout.pages.length===1 ? 'yes' : 'no'} • ${plan.lyricFamily}/${plan.chordFamily}`,
        margin,
        pageH - (margin * 0.6)
      )
    }

    p.columns.forEach((col) => {
      let x = col.x
      let y = contentStartY
      for (const b of col.blocks) {
        if (b.type === 'section') {
          y += sectionTopPad
          doc.setFont(lFam, 'bold'); doc.setFontSize(sectionSize)
          doc.text(`[${b.header}]`, x, y)
          y += sectionSize + 4
        } else if (b.type === 'line') {
          if (b.chords?.length) {
            doc.setFont(cFam, 'bold'); doc.setFontSize(plan.chordSizePt)
            for (const c of b.chords) doc.text(c.sym, x + c.x, y)
            y += plan.chordSizePt + lineGap / 2
          }
          doc.setFont(lFam, 'normal'); doc.setFontSize(plan.lyricSizePt)
          doc.text(b.lyrics, x, y)
          y += plan.lyricSizePt + lineGap
        }
      }
    })
  })
}

/* -----------------------------------------------------------
 * Exposed helpers
 * --------------------------------------------------------- */
export { planSongLayout, chooseBestLayout, normalizeSongInput } from './pdfLayout'

export async function chooseBestLayoutAuto(song, baseOpt = {}) {
  const doc = await newPDF()
  let fams = {}
  try { fams = await ensureFontsEmbedded(doc) } catch {}
  const o = {
    lyricFamily: fams.lyricFamily || baseOpt.lyricFamily || 'Helvetica',
    chordFamily: fams.chordFamily || baseOpt.chordFamily || 'Courier',
    ...baseOpt,
  }
  return planWithDoc(doc, normalizeSongInput(song), o)
}

/* -----------------------------------------------------------
 * Single-song PDF
 * --------------------------------------------------------- */
export async function downloadSingleSongPdf(song, options) {
  const doc = await newPDF()
  let fams = {}
  try { fams = await ensureFontsEmbedded(doc) } catch {}
  const base = {
    lyricSizePt: Math.max(12, options?.lyricSizePt || 16),
    chordSizePt: Math.max(12, options?.chordSizePt || 16),
    margin: 36,
    lyricFamily: fams.lyricFamily || 'Helvetica',
    chordFamily: fams.chordFamily || 'Courier',
    columns: options?.columns === 2 ? 2 : 1,
  }
  const norm = normalizeSongInput(song)
  const { plan } = planWithDoc(doc, norm, base)
  const title = song.title || norm.title || 'Untitled'
  const key = song.key || norm.key || 'C'
  drawPlannedSong(doc, plan, { title, key })
  doc.setProperties({ title: `${title} – GraceChords` })
  doc.save(`${title.replace(/\s+/g, '_')}.pdf`)
  return { plan }
}

/* -----------------------------------------------------------
 * Multi-song PDF (setlists / songbooks)
 * --------------------------------------------------------- */
export async function downloadMultiSongPdf(songs, options = {}) {
  const doc = await newPDF()
  let fams = {}
  try { fams = await ensureFontsEmbedded(doc) } catch {}
  const baseOpt = {
    lyricSizePt: Math.max(12, options?.lyricSizePt || 16),
    chordSizePt: Math.max(12, options?.chordSizePt || 16),
    margin: 36,
    lyricFamily: fams.lyricFamily || 'Helvetica',
    chordFamily: fams.chordFamily || 'Courier',
    columns: options?.columns === 2 ? 2 : 1,
  }
  const includeTOC = !!options?.includeTOC
  const cover = options?.coverImageDataUrl

  const planned = songs.map((s, idx) => {
    const norm = normalizeSongInput(s)
    const { plan } = planWithDoc(doc, norm, baseOpt)
    const title = s.title || norm.title || 'Untitled'
    const key = s.key || norm.key || 'C'
    return {
      raw: s,
      plan,
      title,
      key,
      num: s._songbookNum || idx + 1,
      origTitle: s._origTitle || title,
    }
  })

  const pageW = doc.internal.pageSize.getWidth()
  const pageH = doc.internal.pageSize.getHeight()
  const margin = baseOpt.margin
  const lineH = 16
  const linesPerPage = Math.floor((pageH - margin * 2 - 40) / lineH)
  const tocPages = includeTOC ? Math.ceil(planned.length / linesPerPage) : 0
  let curPage = 1 + (cover ? 1 : 0) + tocPages
  const tocEntries = planned.map((p) => {
    const entry = { num: p.num, title: p.origTitle, page: curPage }
    curPage += p.plan.layout.pages.length
    return entry
  })

  // Cover page
  if (cover) {
    try { doc.addImage(cover, 'JPEG', 0, 0, pageW, pageH, undefined, 'FAST') } catch {}
    if (includeTOC || planned.length) doc.addPage()
  }

  // Table of contents
  if (includeTOC) {
    doc.setFont(baseOpt.lyricFamily, 'bold'); doc.setFontSize(24)
    doc.text('Table of Contents', pageW / 2, margin + 20, { align: 'center' })
    doc.setFont(baseOpt.lyricFamily, 'normal'); doc.setFontSize(12)
    let y = margin + 40
    tocEntries.forEach((e, i) => {
      if (i > 0 && i % linesPerPage === 0) { doc.addPage(); y = margin }
      const leftText = `${e.num} ${e.title}`
      doc.text(leftText, margin, y)
      const pageStr = String(e.page)
      const right = pageW - margin
      const textW = doc.getTextWidth(leftText)
      const pageWdt = doc.getTextWidth(pageStr)
      const dotsW = right - margin - textW - pageWdt - 4
      if (dotsW > 0) {
        const dot = doc.getTextWidth('.')
        const dots = '.'.repeat(Math.max(0, Math.floor(dotsW / dot)))
        doc.text(dots, margin + textW + 2, y)
      }
      doc.text(pageStr, right, y, { align: 'right' })
      y += lineH
    })
    if (planned.length) doc.addPage()
  }

  // Songs
  planned.forEach((p, idx) => {
    if (idx > 0) doc.addPage()
    drawPlannedSong(doc, p.plan, { title: p.title, key: p.key })
  })

  doc.setProperties({ title: options?.docTitle || 'GraceChords Setlist' })
  doc.save(options?.fileName || 'GraceChords_Selection.pdf')
}

/* -----------------------------------------------------------
 * Songbook wrapper
 * --------------------------------------------------------- */
export async function downloadSongbookPdf(songs, { includeTOC, coverImageDataUrl } = {}) {
  const numbered = songs.map((s, i) => ({
    ...s,
    title: `${i + 1}. ${s.title}`,
    _songbookNum: i + 1,
    _origTitle: s.title,
  }))
  const date = new Date().toISOString().slice(0, 10).replace(/-/g, '')
  await downloadMultiSongPdf(numbered, {
    includeTOC,
    coverImageDataUrl,
    docTitle: 'GraceChords Songbook',
    fileName: `songbook-${date}.pdf`,
  })
}
