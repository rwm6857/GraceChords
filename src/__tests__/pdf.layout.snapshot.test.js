// src/__tests__/pdf.layout.snapshot.test.js
import { describe, it, expect } from 'vitest'
import { parseChordPro } from '../utils/chordpro.js'
import { getLayoutMetrics } from '../utils/pdf.js'

const CHORDPRO = `
{title: Test Song}
{key: G}
[VERSE]
[G]This is a [D]line with [Em]chords and [C]lyrics that may wrap around to the next line when wide.
[CHORUS]
[C]Short line
[G]Another short line
[BRIDGE]
[Em]A very very very very very very very very very very very very long line to force two columns
`;

describe('PDF layout snapshot', () => {
  it('chord/lyric positions and columns stay stable (G)', () => {
    const parsed = parseChordPro(CHORDPRO)
    const snap = getLayoutMetrics(parsed, { key: 'G', chordsOn: true })
    expect(snap).toMatchSnapshot()
  })

  it('transposed layout stays stable (A)', () => {
    const parsed = parseChordPro(CHORDPRO)
    const snap = getLayoutMetrics(parsed, { key: 'A', chordsOn: true })
    expect(snap).toMatchSnapshot()
  })
})
