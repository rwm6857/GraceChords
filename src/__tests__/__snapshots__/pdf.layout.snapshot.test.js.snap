// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PDF layout snapshot > chord/lyric positions and columns stay stable (G) 1`] = `
[
  {
    "cols": [
      {
        "blocks": [],
        "c": 0,
      },
    ],
    "p": 0,
  },
]
`;

exports[`PDF layout snapshot > transposed layout stays stable (A) 1`] = `
[
  {
    "cols": [
      {
        "blocks": [],
        "c": 0,
      },
    ],
    "p": 0,
  },
]
`;
